internal/controller/biossettings_controller.go

```diff
@@ -0,0 +1,776 @@
+// +kubebuilder:rbac:groups="batch",resources=jobs,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *BiosSettingsReconciler) Reconcile(
+	ctx context.Context,
+	req ctrl.Request,
+) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	biosSettings := &metalv1alpha1.BiosSettings{}
+	if err := r.Get(ctx, req.NamespacedName, biosSettings); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	log.V(1).Info("Reconciling biosSettings")
+
+	return r.reconcileExists(ctx, log, biosSettings)
+}
+
+// Determine whether reconciliation is required. It's not required if:
+// - object is being deleted;
+// - object does not contain reference to server;
+// - object contains reference to server, but server references to another object with lower version;
+func (r *BiosSettingsReconciler) reconcileExists(
+	ctx context.Context,
+	log logr.Logger,
+	biosSettings *metalv1alpha1.BiosSettings,
+) (ctrl.Result, error) {
+	// if object is being deleted - reconcile deletion
+	if !biosSettings.DeletionTimestamp.IsZero() {
+		log.V(1).Info("object is being deleted")
+		return r.reconcileDeletion(ctx, log, biosSettings)
+	}
+
+	return r.reconcile(ctx, log, biosSettings)
+}
+
+func (r *BiosSettingsReconciler) reconcileDeletion(
+	ctx context.Context,
+	log logr.Logger,
+	biosSettings *metalv1alpha1.BiosSettings,
+) (ctrl.Result, error) {
+	if !controllerutil.ContainsFinalizer(biosSettings, biosSettingsFinalizer) {
+		return ctrl.Result{}, nil
+	}
+	if err := r.cleanupReferences(ctx, log, biosSettings); err != nil {
+		log.Error(err, "failed to cleanup references")
+		return ctrl.Result{}, err
+	}
+	log.V(1).Info("ensured references were cleaned up")
+
+	log.V(1).Info("Ensuring that the finalizer is removed")
+	if modified, err := clientutils.PatchEnsureNoFinalizer(ctx, r.Client, biosSettings, biosSettingsFinalizer); err != nil || modified {
+		return ctrl.Result{}, err
+	}
+
+	log.V(1).Info("biosSettings is deleted")
+	return ctrl.Result{}, nil
+}
+
+func (r *BiosSettingsReconciler) cleanupServerMaintenanceReferences(
+	ctx context.Context,
+	log logr.Logger,
+	biosSettings *metalv1alpha1.BiosSettings,
+) error {
+	if biosSettings.Spec.ServerMaintenanceRef == nil {
+		return nil
+	}
+	// try to get the serverMaintaince created
+	serverMaintenance, err := r.getReferredServerMaintenance(ctx, log, biosSettings.Spec.ServerMaintenanceRef)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return fmt.Errorf("failed to get referred serverMaintenance obj from biosSettings: %w", err)
+	}
+
+	// if we got the server ref. by and its not being deleted
+	if err == nil && serverMaintenance.DeletionTimestamp.IsZero() {
+		// if the serverMaintenace is not created by biosSettings Controller, dont delete.
+		labelsOnMaintenance := serverMaintenance.GetLabels()
+
+		// created by the controller
```

Comment:
We could leverage here `metav1.IsControlledBy()`. That way we don't need to set extra labels on the `Maintenance` object.